### PR TITLE
Prepare for release 3.5.0

### PR DESCRIPTION
**Pull request by marcopeereboom**
@LordReg let's get 'r done.

I created this PR so that we can pile on the changes we want to push into 3.5.0.

### CODE REVIEW

**Review by marcopeereboom — COMMENTED** [review]
I generally agree with the changes. I am questioning the amount of new macros being added. It feels like some should be functions (static and/or inline if people think that may help the compiler). I know it is a fine line between "getting a window id" and "rotate this" but one is inspective and the other is more (but not exactly a function) of a functional thing.

I just want to point this out and have people think for a minute or two about inspection vs function.

**Review by marcopeereboom — COMMENTED** [review]
Actually, I don't like this one. I prefer to strip the binary afterwards. Not having debug symbols is a giant pain in the ass and a waste of time when we get a report from the field. Besides, most os' redo this anyway. I'd like to revert the removal of `-g`

**Comment by marcopeereboom** [comment]
Anything else we want to sneak in?

**Comment by LordReg** [comment]
It is patched out everywhere, including OpenBSD ports. Users filing a report should really build with `-DSWM_DEBUG` as well as `-g`.  Any thoughts about having it be a default for CFLAGS when it is unset?
e.g.
```
CFLAGS ?= -g
CFLAGS +=-std=c99 -Wmissing-prototypes -Wall -Wextra -Wshadow -Wno-uninitialized
```

**Comment by marcopeereboom** [comment]
Our makefiles should be examples and useful for developers. OS bundlers will do whatever they do. Thus, we should make it easier on ourselves and always build with `-g`. If it matters to Arch they'll strip it or not build with `-g`. `-g` is good!

**Comment by LordReg** [comment]
Nearly there.

**Comment by marcopeereboom** [comment]
Going twice!

**Comment by marcopeereboom** [comment]
Guess we’ll start at going once again 😃

File: spectrwm.c [hunk 1]
```diff
@@ -5567,7 +5655,7 @@ focus_region(struct swm_region *r)
 		set_region(r);
 
 		/* New region is empty; need to manually unfocus win. */
-		if (old_r) {
+		if (old_r != r) {
```
Comment (bket) [inline spectrwm.c:5702]: @LordReg, this change causes spectrwm to segfault upon starting on OpenBSD amd64 current:

```
Program terminated with signal SIGSEGV, Segmentation fault.
#0  focus_region (r=0x87ab32ba9c0) at spectrwm.c:5659
5659                            unfocus_win(old_r->ws->focus);
(gdb) bt full
#0  focus_region (r=0x87ab32ba9c0) at spectrwm.c:5659
        none = <error reading variable none (Cannot access memory at address 0x0)>
        nfw = <optimized out>
        old_r = 0x0
#1  0x00000878a0ad6d97 in setup_focus () at spectrwm.c:14610
        r = 0x87ab32ba9c0
        w = <optimized out>
#2  0x00000878a0ad87e8 in main (argc=<optimized out>, argv=<optimized out>) at spectrwm.c:15343
        pfd = {{fd = 0, events = 0, revents = 0}, {fd = 0, events = 0, revents = 0}}
        sact = {__sigaction_u = {__sa_handler = 0x878a0ab3ea0 <sighdlr>,
            __sa_sigaction = 0x878a0ab3ea0 <sighdlr>}, sa_mask = 0, sa_flags = 8}
        stdin_ready = <error reading variable stdin_ready (Cannot access memory at address 0x0)>
        cfile = <optimized out>
        ch = <optimized out>
        evt = <optimized out>
        mne = <optimized out>
        i = <optimized out>
        num_screens = 1
        r = <optimized out>
        num_readable = <optimized out>
```

Reverting this bit fixes the above for me.

Comment (LordReg) [reply spectrwm.c]: Thanks for the report, fixed.

**Comment by marcopeereboom** [comment]
Going twice!!

**Comment by marcopeereboom** [comment]
Going thrice!

If I hear nothing over the next few days I am hitting rebase and merge!

**Comment by bket** [comment]
Using `swap_main` causes spectrwm to segfault:

```
Core was generated by `spectrwm'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  get_focus_prev (ws=0x1f9da6c6030) at spectrwm.c:6561
6561			if (!HIDDEN(w) && w != ws->focus && !win_related(w, ws->focus))
(gdb) bt full
#0  get_focus_prev (ws=0x1f9da6c6030) at spectrwm.c:6561
        w = <optimized out>
#1  swapwin (bp=<optimized out>, r=0x1fa591a5600, 
    args=0x1f7a2b67fe0 <actions+5312>) at spectrwm.c:6405
        sw = <optimized out>
        wl = 0x1f9da6c6088
        w = 0x1fa59167780
        pw = <optimized out>
#2  0x000001f7a2b58c2a in keypress (e=0x1fa59197d50) at spectrwm.c:13270
        replay = <optimized out>
        keysym = <optimized out>
        bp = 0x1fa59181440
        ap = <optimized out>
#3  0x000001f7a2b5f485 in main (argc=<optimized out>, argv=<optimized out>)
    at spectrwm.c:15741
        pfd = {{fd = 3, events = 1, revents = 1}, {fd = 0, events = 1, 
            revents = 0}}
        sact = {__sigaction_u = {__sa_handler = 0x1f7a2b3a380 <sighdlr>, 
            __sa_sigaction = 0x1f7a2b3a380 <sighdlr>}, sa_mask = 0, sa_flags = 8}
        stdin_ready = false
        cfile = <optimized out>
        ch = <optimized out>
        evt = 0x1fa59197d50
        mne = <optimized out>
        i = <optimized out>
        num_screens = 1
        r = <optimized out>
        num_readable = <optimized out>

```

Could be related to daf15b92c81357c009b7a1156ba5e32ad7c6623f.

**Comment by LordReg** [comment]
@bket Already fixed, just hadn't pushed yet.

**Comment by bket** [comment]
> @bket Already fixed, just hadn't pushed yet.

Great! Issue seems resolved as of cc76cd4284c50c6e8020f8f49059d6d88a5dfcfe.

**Comment by LordReg** [comment]
> Going thrice!
> 
> If I hear nothing over the next few days I am hitting rebase and merge!

More on the way :smiley:

**Comment by marcopeereboom** [comment]
What about now?

**Comment by wavexx** [comment]
I've tested the last commits, so I'm reporting here some issues with it (hoping it's useful).
After commit 5ab7ad48ed6902d8eca2c4fb1b065bcabce7396e the window borders seem to ignore border_width for me, so while the gap is there, the actual borders seem to be exactly 1 pixel width.

**Comment by LordReg** [comment]
@wavexx Fixed. Was able to reproduce the issue with border_width > bar_border_width (e.g. leaving it default.)

**Comment by marcopeereboom** [comment]
Cal I hit "rebase" yet?

**Comment by wavexx** [comment]
Another issue spotted: rg_move_next, rg_move_prev no longer swap the next/previous region, but seem just to cycle over them

**Comment by wavexx** [comment]
Additionally, it seems that rg_move_next/prev only breaks if workspace_clamp=1

**Comment by LordReg** [comment]
@wavexx fixed

@marcopeereboom might as well be in master at this point :wink:

**Comment by marcopeereboom** [comment]
🎉🎉🎉

**Comment by wavexx** [comment]
So much goodness all at once :love_you_gesture:

**Comment by marcopeereboom** [comment]
Can someone push the tag? I got distracted yesterday before I completed that.

**Comment by LordReg** [comment]
Np, will take care of it once i finish updating the changelog.

**Comment by LordReg** [comment]
There we go 🎆🎉🎊🎆

**Comment by marcopeereboom** [comment]
🎉🍾

**Pull request merged**